cmd/agent/daemon/enrichment/enrichers.go

```diff
@@ -0,0 +1,149 @@
+package enrichment
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"path/filepath"
+	"strconv"
+	"syscall"
+
+	castpb "github.com/castai/kvisor/api/v1/runtime"
+	"github.com/castai/kvisor/pkg/containers"
+	"github.com/castai/kvisor/pkg/logging"
+	lru "github.com/hashicorp/golang-lru/v2"
+	"github.com/minio/sha256-simd"
+)
+
+type fileHashCacheKey string
+type ContainerForCgroupGetter func(cgroup uint64) (*containers.Container, bool, error)
+
+func EnrichWithFileHash(log *logging.Logger, containerForCgroupGetter ContainerForCgroupGetter, procFS fs.StatFS) EventEnricher {
+	cache, err := lru.New[fileHashCacheKey, []byte](1024)
+	if err != nil {
+		// This case can never happen, as err is only thrown if cache size is <0, which it isn't.
+		panic(err)
+	}
+
+	return &fileHashEnricher{
+		log:                      log,
+		containerForCgroupGetter: containerForCgroupGetter,
+		procFS:                   procFS,
+		cache:                    cache,
+	}
+}
+
+type fileHashEnricher struct {
+	log                      *logging.Logger
+	containerForCgroupGetter ContainerForCgroupGetter
+	procFS                   fs.StatFS
+	cache                    *lru.Cache[fileHashCacheKey, []byte]
+}
+
+func (enricher *fileHashEnricher) Enrich(ctx context.Context, e *castpb.Event) {
+	exec := e.GetExec()
+	if exec == nil || exec.Path == "" {
+		return
+	}
+
+	cont, found, err := enricher.containerForCgroupGetter(e.CgroupId)
+	// if we do not have the container cached, we cannot easily accesss the FS and hence simply skip
+	if err != nil || !found {
+		return
+	}
+
+	for _, pid := range cont.PIDs {
+		sha, err := enricher.calcFileHashForPID(cont, pid, exec.Path)
+		// We search for the first PID we can successfully calculate a filehash for.
+		if err != nil {
+			if errors.Is(err, ErrFileDoesNotExist) {
+				// If the wanted file does not exist in the PID mount namespace, it will also not exist in the mounts of the other.
+				// We can hence simply return, as we will not find the wanted file.
+				return
+			}
+
+			continue
+		}
+
+		if exec.Meta == nil {
+			exec.Meta = &castpb.ExecMetadata{}
+		}
+
+		exec.Meta.HashSha256 = sha
+		return
+	}
+}
+
+func (enricher *fileHashEnricher) calcFileHashForPID(cont *containers.Container, pid uint32, execPath string) ([]byte, error) {
+	pidString := strconv.FormatInt(int64(pid), 10)
+
+	_, err := enricher.procFS.Stat(pidString)
```

Comment:
These 2 additional stat syscalls looks redundant. If file doesn't exist you will get error while trying to open it.

Reply:
The problem is that is use this as part of the optimization when checking for files. The error we return here is a different one than checking for the file itself, as if we have a proc folder, but cannot find the file, we can just exit. Otherwise we would check the next PID. 

I do not know if the PIDs we get back from containerD are always ordered, so that the first in the list is PID 1 in the PIDNS. Thinking about it, I believe we can anyway always take the lowest PID, as it should be PID 1. Is there any case you can think of, where a container is running, when PID 1 died?